File: vision_core/src/fmap_loader.py
"""
XNav FMap Loader
Parses WPILib .fmap (field map) JSON files containing AprilTag 3D poses.
"""

import json
import os
import math
import logging
from dataclasses import dataclass, field
from typing import Dict, Optional, List

logger = logging.getLogger(__name__)


@dataclass
class TagPose:
    id: int
    x: float          # meters, field-centric
    y: float
    z: float
    qw: float = 1.0   # quaternion
    qx: float = 0.0
    qy: float = 0.0
    qz: float = 0.0
    # Euler (derived)
    roll: float = 0.0
    pitch: float = 0.0
    yaw: float = 0.0


@dataclass
class FieldMap:
    length: float = 0.0     # meters
    width: float = 0.0
    tags: Dict[int, TagPose] = field(default_factory=dict)


def _quat_to_euler(w, x, y, z):
    """Convert quaternion to (roll, pitch, yaw) in degrees."""
    sinr_cosp = 2 * (w * x + y * z)
    cosr_cosp = 1 - 2 * (x * x + y * y)
    roll = math.atan2(sinr_cosp, cosr_cosp)

    sinp = 2 * (w * y - z * x)
    if abs(sinp) >= 1:
        pitch = math.copysign(math.pi / 2, sinp)
    else:
        pitch = math.asin(sinp)

    siny_cosp = 2 * (w * z + x * y)
    cosy_cosp = 1 - 2 * (y * y + z * z)
    yaw = math.atan2(siny_cosp, cosy_cosp)

    return math.degrees(roll), math.degrees(pitch), math.degrees(yaw)


def load_fmap(path: str) -> Optional[FieldMap]:
    """Load a WPILib .fmap file and return a FieldMap."""
    if not os.path.exists(path):
        logger.warning("FMap file not found: %s", path)
        return None

    try:
        with open(path, "r") as f:
            data = json.load(f)
    except Exception as e:
        logger.error("Failed to parse fmap %s: %s", path, e)
        return None

    fm = FieldMap()

    # Field dimensions (optional in some fmap formats)
    field_info = data.get("field", {})
    fm.length = field_info.get("length", 0.0)
    fm.width = field_info.get("width", 0.0)

    tags_data = data.get("tags", [])
    if not tags_data:
        # Some fmap files use "fiducials"
        tags_data = data.get("fiducials", [])

    for tag_data in tags_data:
        tag_id = tag_data.get("ID") or tag_data.get("id") or tag_data.get("fiducialId")
        if tag_id is None:
            continue
        tag_id = int(tag_id)

        pose = tag_data.get("pose", {})
        translation = pose.get("translation", {})
        tx = float(translation.get("x", 0))
        ty = float(translation.get("y", 0))
        tz = float(translation.get("z", 0))

        rotation = pose.get("rotation", {})
        quat = rotation.get("quaternion", {})
        qw = float(quat.get("W", 1))
        qx = float(quat.get("X", 0))
        qy = float(quat.get("Y", 0))
        qz = float(quat.get("Z", 0))

        roll, pitch, yaw = _quat_to_euler(qw, qx, qy, qz)

        tp = TagPose(
            id=tag_id,
            x=tx, y=ty, z=tz,
            qw=qw, qx=qx, qy=qy, qz=qz,
            roll=roll, pitch=pitch, yaw=yaw
        )
        fm.tags[tag_id] = tp

    logger.info("Loaded fmap with %d tags (field %gx%gm)", len(fm.tags), fm.length, fm.width)
    return fm

File: vision_core_cpp/src/fmap_loader.cpp
#include "fmap_loader.hpp"

#include <fstream>
#include <cmath>
#include <syslog.h>
#include <filesystem>

namespace fs = std::filesystem;

static std::tuple<double,double,double> quatToEuler(double w, double x, double y, double z) {
    // Roll (x-axis)
    double sinr = 2.0 * (w*x + y*z);
    double cosr = 1.0 - 2.0*(x*x + y*y);
    double roll = std::atan2(sinr, cosr);

    // Pitch (y-axis)
    double sinp = 2.0 * (w*y - z*x);
    double pitch = std::abs(sinp) >= 1.0
                 ? std::copysign(M_PI/2.0, sinp)
                 : std::asin(sinp);

    // Yaw (z-axis)
    double siny = 2.0 * (w*z + x*y);
    double cosy = 1.0 - 2.0*(y*y + z*z);
    double yaw = std::atan2(siny, cosy);

    const double r2d = 180.0 / M_PI;
    return {roll*r2d, pitch*r2d, yaw*r2d};
}

std::optional<FieldMap> loadFmap(const std::string& path) {
    if (!fs::exists(path)) {
        syslog(LOG_WARNING, "FMap file not found: %s", path.c_str());
        return std::nullopt;
    }

    json data;
    try {
        std::ifstream f(path);
        data = json::parse(f);
    } catch (const std::exception& e) {
        syslog(LOG_ERR, "Failed to parse fmap %s: %s", path.c_str(), e.what());
        return std::nullopt;
    }

    FieldMap fm;
    if (data.contains("field")) {
        auto& field = data["field"];
        fm.length = field.value("length", 0.0);
        fm.width  = field.value("width",  0.0);
    }

    // Support both "tags" and "fiducials" keys
    json tags_data = json::array();
    if (data.contains("tags"))      tags_data = data["tags"];
    else if (data.contains("fiducials")) tags_data = data["fiducials"];

    for (auto& tag_data : tags_data) {
        int tag_id = -1;
        if (tag_data.contains("ID"))         tag_id = tag_data["ID"].get<int>();
        else if (tag_data.contains("id"))    tag_id = tag_data["id"].get<int>();
        else if (tag_data.contains("fiducialId")) tag_id = tag_data["fiducialId"].get<int>();
        if (tag_id < 0) continue;

        auto& pose = tag_data["pose"];
        auto& trans = pose["translation"];
        auto& rot   = pose["rotation"];
        auto& quat  = rot["quaternion"];

        TagPose tp;
        tp.id = tag_id;
        tp.x  = trans.value("x", 0.0);
        tp.y  = trans.value("y", 0.0);
        tp.z  = trans.value("z", 0.0);
        tp.qw = quat.value("W", 1.0);
        tp.qx = quat.value("X", 0.0);
        tp.qy = quat.value("Y", 0.0);
        tp.qz = quat.value("Z", 0.0);

        auto [roll, pitch, yaw] = quatToEuler(tp.qw, tp.qx, tp.qy, tp.qz);
        tp.roll = roll; tp.pitch = pitch; tp.yaw = yaw;

        fm.tags[tag_id] = tp;
    }

    syslog(LOG_INFO, "Loaded fmap with %zu tags (field %.1fx%.1fm)",
           fm.tags.size(), fm.length, fm.width);
    return fm;
}
